README
Perl::PrereqScanner::Scanner::Runtime

This distribution contains a collection of various run-time scanners,
intended to expose "optional" prerequisites to tools that are expressly
looking for them.

Grading them as optional is presently an exercise left to the user.

This class is a convenience wrapper around a bunch of other classes that
provide different kinds of generic "Runtime" detection.

*   "Module::Runtime"

    Scans for "Module::Runtime" specific syntax.

*   "Class::Load"

    Scans for "Class::Load" specific syntax.

*   "Module::Load"

    Scans for "Module::Load" specific syntax.

  AUDIENCE

Its worth noting that this module is not necessarily going to be useful
in part of a "Dist::Zilla" or similar stack. By nature of how this
works, its job is to find dependencies that likely qualify as
*optional*.

Hence, blindly advertising what this scanner reports as dependencies is
not recommended.

This tool is more oriented around turn-key maintainers who want a quick
way of finding *possible* auto-magical dependency behavior so they can
make judgment calls about it.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Perl-PrereqScanner-Scanner-Runtime

0.001000 2016-04-09T18:05:02Z 71219d7
 - First version.



lib/Perl/PrereqScanner/Scanner/Class/Load.pm
use 5.006;    # our
use strict;
use warnings;

package Perl::PrereqScanner::Scanner::Class::Load;

our $VERSION = '0.001000';

# ABSTRACT: Extract Class::Load module requirements

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moose qw( with );
use Carp qw( croak );
with 'Perl::PrereqScanner::Scanner';

__PACKAGE__->meta->make_immutable;
no Moose;

my %functions = (
  load_class          => '_get_pn',
  try_load_class      => '_get_pn',
  load_optional_class => '_get_pn',
);
my %subs = ( map { ( "Class::Load::$_" => [ $functions{$_}, $_ ], "$_" => [ $functions{$_}, $_ ], ) } keys %functions );





sub scan_for_prereqs {
  my ( $self, $ppi_doc, $req ) = @_;

  # Compiles a list of function names to grep for
  my $literal_re = sprintf q[\A(?:%s)\z], join q[|], map { quotemeta }
    sort { $a cmp $b } _using_class_load($ppi_doc),
    map  { 'Class::Load::' . $_ } keys %functions;

  # Scan document for <$functionname $quoted thing,...> or
  #   <$functionname(@arglist)> statements
  #
  # Skimming all statements for C<$functionname> and the looking right to see what to do
  my (@interesting);
  $ppi_doc->find(
    sub {
      return q[] unless $_[1]->isa('PPI::Statement');
      my (@children) = $_[1]->schildren;
      my ( $child, $previous );
      while (@children) {
        $previous = $child;
        $child    = shift @children;

        # Match sub call
        next
          unless $child->isa('PPI::Token::Word')
          and $child->literal =~ qr/$literal_re/sx;

        # Skip methods
        next if $previous and $previous->isa('PPI::Token::Operator') and q{->} eq $previous->content;

        # Handle a list of arguments as token->next
        if ( $children[0]->isa('PPI::Structure::List') ) {
          push @interesting, [ $child->literal, shift @children ];
          next;
        }

        # Handle "bare" calls by passing *all* remaining siblings
        # if token-next is a module name
        if ( $children[0]->isa('PPI::Token::Quote')
          or $children[0]->isa('PPI::Token::Word')
          or $children[0]->isa('PPI::Token::QuoteLike::Words') )
        {
          push @interesting, [ $child->literal, @children ];
          next;
        }

        # Here things fall through.
      }
      return q[];
    },
  );

  # Processing the identified groups happens independently to avoid
  # repressed errors, but this logic could just as easily be internal
  for my $group (@interesting) {
    my $subname = shift @{$group};
    my $handler = $self->can( $subs{$subname}->[0] )
      or croak "Can't find extractor for $subname";
    my ( $module, $version ) = $handler->($group);
    next unless defined $module;
    local $@ = undef;
    ## no critic (RequireCheckingReturnValueOfEval)
    eval { $req->add_minimum( $module, 0 ) };    # If module name is invalid, not adding it is ok
    eval { $req->add_string_requirement( $module, $version ) };    # if version is invalid, not adding it is ok
  }
  return 1;
}

# Tries to flatten series of tokens/nodes to strings.
# may have bugs...
# anything that isn't known-safely stringifiable says a ref
# list seperators are nommed
sub _flatten {    ## no critic (RequireArgUnpacking)
  my (@children) = @{ $_[0] };
  my (@out);
  while ( my $child = shift @children ) {
    next
      if $child->isa('PPI::Token::Operator')
      and $child->content =~ qr{,|=>}sx;
    if ( $child->isa('PPI::Token::Quote::Single')
      or $child->isa('PPI::Token::Quote::Literal') )
    {
      push @out, $child->string;
      next;
    }
    if ( $child->isa('PPI::Token::Quote::Double') ) {
      if ( not $child->interpolations ) {
        push @out, $child->string;
        next;
      }
      push @out, $child;
      next;
    }
    if ( $child->isa('PPI::Token::Word') ) {
      if ( $children[0]->isa('PPI::Token::Operator')
        and q{=>} eq $children[0]->content )
      {
        shift @children;
        push @out, $child->literal;
        next;
      }
      push @out, $child;
      next;
    }
    if ( $child->isa('PPI::Token::Number')
      or $child->isa('PPI::Token::QuoteLike::Words') )
    {
      push @out, $child->literal;
      next;
    }
    if ( $child->isa('PPI::Structure::List')
      or $child->isa('PPI::Statement::Expression') )
    {
      push @out, _flatten( [ $child->schildren ] );
      next;
    }
    push @out, $child;    # Stash unhandled tokens verbatim
  }
  return @out;
}

sub _get_pn {
  my ($schildren) = @_;
  my (@args)      = _flatten($schildren);
  if ( defined $args[0] and not ref $args[0] ) {
    return ( $args[0], 0 );
  }
  return;
}

sub _get_pn_v {
  my ($schildren) = @_;
  my (@args)      = _flatten($schildren);
  if ( defined $args[0] and not ref $args[0] ) {
    if ( defined $args[1] and not ref $args[1] ) {
      return ( $args[0], $args[1] );
    }
    return ( $args[0], 0 );
  }
  return;
}

sub _using_class_load {
  my ($doc) = @_;
  my $found = undef;
  $doc->find_first(
    sub {
      return q[]
        if not $_[1]->isa('PPI::Statement::Include')
        or 'use' ne $_[1]->type
        or $_[1]->pragma
        or 'Class::Load' ne $_[1]->module;
      return $found = 1;
    },
  );

  # TODO: Detect all functions imported instead of assuming any
  # 'use' anywhere implies everything.
  return keys %functions if $found;
  return;
}




















1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Perl::PrereqScanner::Scanner::Class::Load - Extract Class::Load module requirements

=head1 VERSION

version 0.001000

=head1 DESCRIPTION

This module is an extension for C<Perl::PrereqScanner> which extracts the following C<Class::Load>
syntax and declares them as requirements.

  use Class::Load;

  load_class("Example1");
  try_load_class("Example2");
  load_optional_class("Example3");

  Class::Load::load_class("Example1");
  Class::Load::try_load_class("Example2");
  Class::Load::load_optional_class("Example3");

B<NOTE>: Version extraction presently unsupported.

=for Pod::Coverage scan_for_prereqs

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Perl/PrereqScanner/Scanner/Module/Runtime.pm
use 5.006;    # our
use strict;
use warnings;

package Perl::PrereqScanner::Scanner::Module::Runtime;

our $VERSION = '0.001000';

# ABSTRACT: Extract Module::Runtime module requirements

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moose qw( with );
use Carp qw( croak );
with 'Perl::PrereqScanner::Scanner';

__PACKAGE__->meta->make_immutable;
no Moose;

my %functions = (
  require_module             => '_get_pn',
  use_module                 => '_get_pn_v',
  use_package_optimistically => '_get_pn_v',
);
my %subs = ( map { ( "Module::Runtime::$_" => [ $functions{$_}, $_ ], "$_" => [ $functions{$_}, $_ ], ) } keys %functions );





sub scan_for_prereqs {
  my ( $self, $ppi_doc, $req ) = @_;

  # Compiles a list of function names to grep for
  my $literal_re = sprintf q[\A(?:%s)\z], join q[|], map { quotemeta }
    sort { $a cmp $b } _using_module_runtime($ppi_doc), map { 'Module::Runtime::' . $_ } keys %functions;

  # Scan document for <$functionname $quoted thing,...> or
  #   <$functionname(@arglist)> statements
  #
  # Skimming all statements for C<$functionname> and the looking right to see what to do
  my (@interesting);
  $ppi_doc->find(
    sub {
      return q[] unless $_[1]->isa('PPI::Statement');
      my (@children) = $_[1]->schildren;
      my ( $child, $previous );
      while (@children) {
        $previous = $child;
        $child    = shift @children;

        # Match sub call
        next unless $child->isa('PPI::Token::Word') and $child->literal =~ qr/$literal_re/sx;

        # Skip methods
        next if $previous and $previous->isa('PPI::Token::Operator') and q{->} eq $previous->content;

        # Handle a list of arguments as token->next
        if ( $children[0]->isa('PPI::Structure::List') ) {
          push @interesting, [ $child->literal, shift @children ];
          next;
        }

        # Handle "bare" calls by passing *all* remaining siblings
        # if token-next is a module name
        if ( $children[0]->isa('PPI::Token::Quote')
          or $children[0]->isa('PPI::Token::Word')
          or $children[0]->isa('PPI::Token::QuoteLike::Words') )
        {
          push @interesting, [ $child->literal, @children ];
          next;
        }

        # Here things fall through.
      }
      return q[];
    },
  );

  # Processing the identified groups happens independently to avoid
  # repressed errors, but this logic could just as easily be internal
  for my $group (@interesting) {
    my $subname = shift @{$group};
    my $handler = $self->can( $subs{$subname}->[0] ) or croak "Can't find extractor for $subname";
    my ( $module, $version ) = $handler->($group);
    next unless defined $module;
    local $@ = undef;
    ## no critic (RequireCheckingReturnValueOfEval)
    eval { $req->add_minimum( $module, 0 ) };    # If module name is invalid, not adding it is ok
    eval { $req->add_string_requirement( $module, $version ) };    # if version is invalid, not adding it is ok
  }
  return 1;
}

# Tries to flatten series of tokens/nodes to strings.
# may have bugs...
# anything that isn't known-safely stringifiable says a ref
# list seperators are nommed
sub _flatten {    ## no critic (RequireArgUnpacking)
  my (@children) = @{ $_[0] };
  my (@out);
  while ( my $child = shift @children ) {
    next if $child->isa('PPI::Token::Operator') and $child->content =~ qr{,|=>}sx;
    if ( $child->isa('PPI::Token::Quote::Single')
      or $child->isa('PPI::Token::Quote::Literal') )
    {
      push @out, $child->string;
      next;
    }
    if ( $child->isa('PPI::Token::Quote::Double') ) {
      if ( not $child->interpolations ) {
        push @out, $child->string;
        next;
      }
      push @out, $child;
      next;
    }
    if ( $child->isa('PPI::Token::Word') ) {
      if ( $children[0]->isa('PPI::Token::Operator') and q{=>} eq $children[0]->content ) {
        shift @children;
        push @out, $child->literal;
        next;
      }
      push @out, $child;
      next;
    }
    if ( $child->isa('PPI::Token::Number')
      or $child->isa('PPI::Token::QuoteLike::Words') )
    {
      push @out, $child->literal;
      next;
    }
    if ( $child->isa('PPI::Structure::List') or $child->isa('PPI::Statement::Expression') ) {
      push @out, _flatten( [ $child->schildren ] );
      next;
    }
    push @out, $child;    # Stash unhandled tokens verbatim
  }
  return @out;
}

sub _get_pn {
  my ($schildren) = @_;
  my (@args)      = _flatten($schildren);
  if ( defined $args[0] and not ref $args[0] ) {
    return ( $args[0], 0 );
  }
  return;
}

sub _get_pn_v {
  my ($schildren) = @_;
  my (@args)      = _flatten($schildren);
  if ( defined $args[0] and not ref $args[0] ) {
    if ( defined $args[1] and not ref $args[1] ) {
      return ( $args[0], $args[1] );
    }
    return ( $args[0], 0 );
  }
  return;
}

sub _using_module_runtime {
  my ($doc) = @_;
  my $found = undef;
  $doc->find_first(
    sub {
      return q[]
        if not $_[1]->isa('PPI::Statement::Include')
        or 'use' ne $_[1]->type
        or $_[1]->pragma
        or 'Module::Runtime' ne $_[1]->module;
      return $found = 1;
    },
  );

  # TODO: Detect all functions imported instead of assuming any
  # 'use' anywhere implies everything.
  return keys %functions if $found;
  return;
}























1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Perl::PrereqScanner::Scanner::Module::Runtime - Extract Module::Runtime module requirements

=head1 VERSION

version 0.001000

=head1 DESCRIPTION

This module is an extension for C<Perl::PrereqScanner> which extracts the following C<Module::Runtime>
syntax and declares them as requirements.

  use Module::Runtime;

  require_module("Example1");
  use_module("Example2");
  use_module("Example3",5.6);
  use_package_optimistically("Example4");
  use_package_optimistically("Example5", 5.6);


  Module::Runtime::require_module("Example1");
  Module::Runtime::use_module("Example2");
  Module::Runtime::use_module("Example3",5.6);
  Module::Runtime::use_package_optimistically("Example4");
  Module::Runtime::use_package_optimistically("Example5", 5.6);

=for Pod::Coverage scan_for_prereqs

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.043.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Detect various kinds of run-time prerequisites",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Perl-PrereqScanner-Scanner-Runtime",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Perl::PrereqScanner::Scanner::Runtime",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Moose" => 0,
    "Perl::PrereqScanner::Scanner" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Perl::PrereqScanner" => 0,
    "Test::More" => "0.89"
  },
  "VERSION" => "0.001000",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t t/class-load/*.t t/module-runtime/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moose" => 0,
  "Perl::PrereqScanner" => 0,
  "Perl::PrereqScanner::Scanner" => 0,
  "Test::More" => "0.89",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
